constantine/math_compiler/codegen_nvidia.nim

```diff
@@ -215,3 +218,321 @@ proc exec*[T](jitFn: CUfunction, r: var T, a, b: T) =
   check cuMemFree(rGPU)
   check cuMemFree(aGPU)
   check cuMemFree(bGPU)
+
+proc getTypes(n: NimNode): seq[NimNode] =
+  case n.kind
+  of nnkIdent, nnkSym: result.add getTypeInst(n)
+  of nnkBracket, nnkTupleConstr, nnkPar:
+    for el in n:
+      result.add getTypes(el)
+  else:
+    error("Arguments to `execCuda` must be given as a bracket or tuple. Instead: " & $n.treerepr)
+
+proc requiresCopy(n: NimNode): bool =
+  ## Returns `true` if the given type is not a trivial data type, which implies
+  ## it will require copying its value manually.
+  case n.typeKind
+  of ntyBool, ntyChar, ntyInt .. ntyUint64: # range includes all floats
+    result = false
+  else:
+    result = true
+
+proc getIdent(n: NimNode): NimNode =
+  ## Generate a `GPU` suffixed ident
+  doAssert n.kind in {nnkIdent, nnkSym}, "No, was: " & $n.treerepr
+  result = ident(n.strVal & "GPU")
+
+proc determineDevicePtrs(r, i: NimNode, iTypes: seq[NimNode]): seq[(NimNode, NimNode)] =
+  ## Returns the device pointer ident and its associated original symbol.
+  for el in r:
+    #if not el.requiresCopy:
```

Comment:
I had some case where I triggered the error when I didn't expect to. Wanted to investigate at a later time and forgot and now I can't reproduce it. Will have another look and otherwise put it back in.

Reply:
Remembered the use case where it caused issues. I've now added test cases for `execCuda` and changed the logic to a `allowedCopy` for this. I.e. if one passes a `var` symbol as a result parameter, it is considered allowed (despite not being a `ptr T` type potentially). The macro logic now checks if it needs to apply an `addr` to the argument or not.